netket/graph/lattice.py

```diff
@@ -20,14 +20,67 @@
 import networkx as _nx
 import warnings
 from typing import Tuple, Union, Optional
+from math import pi
+from itertools import product
+from dataclasses import dataclass
 
-cutoff_tol = 1e-5
+from netket.utils.semigroup import Identity, Element
+from .symmetry import SymmGroup
+
+tol_digits = 5
+cutoff_tol = _np.power(10.0, -tol_digits)
 """Tolerance for the maximum distance cutoff when computing the sparse distance matrix.
 This is necessary because of floating-point errors when computing the distance in non-trivial 
 lattices.
 """
 
 
+@dataclass(frozen=True)
+class Translation(Element):
+    perms: Tuple[Tuple[int]]
+    shift: Tuple[int]
+
+    def __call__(self, sites):
+        for i, dim in enumerate(self.shift):
+            perm = self.perms[i]
+            for j in range(dim):
+                sites = _np.take(sites, perm)
+
+        return sites
+
+    def __repr__(self):
+        return f"T{self.shift}"
+
+
+@dataclass(frozen=True)
+class PlanarRotation(Element):
+    perm: Tuple[int]
+    num_rots: int
+
+    def __call__(self, sites):
+        for i in range(self.num_rots):
+            sites = _np.take(sites, self.perm)
+
+        return sites
+
+    def __repr__(self):
+        return f"Rot{self.num_rots}"
+
+
+@dataclass(frozen=True)
+class Reflection(Element):
+
+    perm: Tuple[int]
+
+    def __call__(self, sites):
+        sites = _np.take(sites, self.perm)
+
+        return sites
+
+    def __repr__(self):
+        return f"Ref"
+
+
```

Comment:
Do I see it correctly that these classes are just "rebranded" permutations? In that case, it might be better to use `netket.utils.semigroup.Permutation` (which already implements this).

Maybe it makes sense to subclass the `Permutation` class so that it is immediately visible what kinds of operations a specific element consists of. But I would prefer not duplicate the actual code that applies the permutation.

Reply:
I think the trick is that `__call__` applies the same permutation a number of times, so you can generate all elements of the rotation group (say) from a single rotation. I'm not sure if `Permutation` is happy with this implementation

Reply:
Since a product of permutations is still a permutation, that amounts to specifying the final permutation indices once on creation of the element, right? (One could even construct the permutations and use `product()` repeatedly, which is implemented for the types `Permutation × Permutation -> Permutation`.

Reply:
That is true. I'm not sure though whether polishing this code is the best way of spending our efforts, given that implementing #703 will tear all of this up...

Reply:
Btw, I tried to understand the code of `SymmGroup::__product_table` and I'm a bit puzzled by the fact that you can apparently use hashes to compare very different objects, like `semigroup::Identity` and an identity permutation.

Reply:
The problem is not whether we have an `Identity` element or not.

First, hashes should never be used alone to check for equality of elements; distinct objects can have the same hash, only the converse implication (equal objects have equal hashes) is required.

Second, equivalent group elements (like `Identity` and the identity permutation or a `Translation` by zero etc.) may compare as unequal. This was a deliberate decision to reduce complexity in the implementation so far. Elements of `SymmGroup` (which is tied to a specific graph) should only be compared based on their action on the sites (i.e., whether `e1(nodes)` and `e2(nodes)` are equal). If `__product_table` does not do this, it should be fixed.

Reply:
@attila-i-szabo It seems to me `__product_table` works on `automorphisms = self.to_array()`, which is an array of permutations in all cases.

There might be a problem with the hashing code (even for the permutation indices) since, as mentioned, unequal elements may in principle have the same hash (though this is rare in practice which could make an error here unlikely to spot in a test).

Reply:
> and I'm a bit puzzled by the fact that you can apparently use hashes to compare very different objects, like semigroup::Identity and an identity permutation

There’s no way this is happening/would work. The hashes only compare the array output of SymmGroup which is always a permutation (for Identity as well). 

>Identity sounds like a dangerous object to have!

I think the Identity element makes things a lot nicer, and I don’t get why it would be dangerous.

Reply:
> There might still be a problem with the hashing code (even for the permutation indices) since, as mentioned, unequal elements may in principle have the same hash (though this is rare in practice which could make an error here unlikely to spot in a test).

I think I can live with this possibility of error, it's highly unlikely any practical case would error out on this...

> @attila-i-szabo It seems to me `__product_table` works on `automorphisms = self.to_array()`, which is an array of permutations in all cases.

That's good, I haven't spotted that. 

However, I would change this going forward. You mentioned a while ago that `PointGroup` in #703 could be given the same product table, etc. routines as `SymmGroup` has. It turns out this would be quite useful – I found a very simple algorithm to generate character tables for the point groups from the times table, which would be simple enough to code that we could expose it to the users. But then, we would replicate a lot of logic between different specific groups, which we might want to move over to `SemiGroup` or some appropriate subclass. However, for this to work generally with a hash-based test, we can't have a jolly-joker `Identity` element...

Reply:
@femtobit I agree with that idea but remember permutation forces the label to be Permutation() which we don’t want. I could change permutation to take a user specified label perhaps?

Reply:
> However, for this to work generally with a hash-based test,

I am not sure what you mean. There is no possibility of working solely with a "hash-based" test in Python, one always has to equality-compare if the hashes match.

If you argue that all elements of a `Semigroup` should have the same type (is this the issue you have with the identity object?), that would be a very strong restriction: Even without the `Identity`, `Translation`s, `Rotation`s, and their mixed products (which have type `Composite`) all have different types.

Reply:
> I agree with that idea but remember permutation forces the label to be Permutation() which we don’t want. I could change permutation to take a user specified label perhaps?

Maybe subclassing `Permutation` is an option? Even something like
```
class Translation(Permutation):
    pass
```
might be enough for the labeling. (Though this will "decay" to a `Permutatrion` as soon as you take a product or similar unless you provide a custom `product` method as well.)

Reply:
> > However, for this to work generally with a hash-based test,
> 
> I am not sure what you mean. There is no possibility of working solely with a "hash-based" test in Python (unless of course the objects you are working with adhere to a stronger contract than the usual Python convention, but that's not easy to achieve in the general case -- also, a one-to-one mapping between objects and hash values is no longer "hashing" in the sense of the word [well, there is "perfect hashing", but this line of thought is getting to theoretical maybe for what amounts to a question of naming], its essentially "encoding" then), one always has to equality-compare if the hashes match.

Sure, but I am pragmatic about this – it's highly unlikely any solely hash-based test would return a wrong result. We can of course run a test on the hashtables we build to make sure they have as many entries as we expect, but I wouldn't slow code down with actual equality testing.

> If you argue that all elements of a `Semigroup` should have the same type (is this the issue you have with the identity object?), that would be a very strong restriction: Even without the `Identity`, `Translation`s, `Rotation`s, and their mixed products (which have type `Composite`) all have different types.

OK, I'll have to think about this a bit more. I'm not terribly worried about `Translation`s and `Rotation`s being a different type, because we only ever multiply them as permutations of a graph. Maybe we can have a `Group` class that is required to produce a canonical representation of its elements that can be compared directly (with or without a hash in between)

Reply:
Why would it break if we overwrite the __repr__ function?

Reply:
@chrisrothUT It wouldn't break, just that `product(SubclassOfPermutation, SubclassOfPermutation)` would be of the (plain) type `Permutation`. OTOH, the current code I think would produce `product(lattice.Translation, lattice.Translation) -> Composite`, which prints as `T(...) @ T(...)`.

Reply:
> Sure, but I am pragmatic about this – it's highly unlikely any solely hash-based test would return a wrong result.

@attila-i-szabo I'd be a bit careful with that in Python, for example `hash(-1) == hash(-2) == -2` and if we have elements containing tuples, it's not that hard to imagine a case where, say, `(1, -1, 3)` and `(1, -2, 3)` could be hashed (which of course also collide).

But, anyways, this discussion is mostly academic. Comparing elements by hash and only comparing the elements if the hashes happen to match is easy to do and not even a performance problem, really. (And I do argue we should do that if we care about the correctness of the result.)

Reply:
> @attila-i-szabo I'd be a bit careful with that in Python, for example `hash(-1) == hash(-2) == -2` and if we have elements containing tuples, it's not that hard to imagine a case where, say, `(1, -1, 3)` and `(1, -2, 3)` (which of course also collide) represent distinct group elements.

@femtobit Ouch, I would've expected better from python. The good thing is that we usually compare integer arrays and the other 8 bytes help stir things up a lot.
 
> But, anyways, this discussion is mostly academic. Comparing elements by hash and only comparing the elements if the hashes happen to match is easy to do and not even a performance problem, really. (And I do argue we should do that if we care about the correctness of the result.)

The problem is not so much direct equality testing (I wouldn't do that via a hash), but matching large amounts of stuff quickly (so as to avoid N^2 algorithms that @gcarleo doesn't like, nor do I). For that, hash collisions would be *very* annoying

Reply:
> The problem is not so much direct equality testing (I wouldn't do that via a hash), but matching large amounts of stuff quickly

If that is about the `hash_coordinates`: If the coordinates themselves can be both hashed and equality-compare (you are rounding the before you compare them, from what I am seeing, right?), why use a `dict` of the form `{hash: position}`? Python `dict`s are using the hash of the key for lookup and equality compare if there is a collision.

Thus, can't we just have
```
self._hash_positions = {
    HashableArray(elements): index
    for index, element in enumerate(int_positions)
}        
```
(or use `elements.tobytes()` as key, if you want, both should work) instead of using `hash(elements.tobytes())` as key?
That should have essentially the same lookup efficiency (`dict` hashes the key anyways) and is more likely to be correct in edge cases.

Reply:
> If that is about the `hash_coordinates`: 

It's a more general pattern: we also use hash-based matching in `__product_table`, which might proliferate further.

> If the coordinates themselves can be both hashed and equality-compared properly (you are rounding them before you compare them, from what I am seeing, right?), why use a `dict` of the form `{hash: index}`? Python `dict`s are already using the hash of the key for lookup and properly handle hash collisions if they occur.
> 
> Thus, can't we just have
> 
> ```
> self._hash_positions = {
>     HashableArray(elements): index
>     for index, element in enumerate(int_positions)
> }        
> ```
> 
> (or use `elements.tobytes()` as key, if you want, both should work) instead of using `hash(elements.tobytes())` as key?
> That should have essentially the same lookup efficiency (`dict` hashes the key anyways) and is more likely to be correct in edge cases.

This is great! I wasn't sure about the efficiency of using a `dict` with complicated keys, but if it works well, then this is way cleaner than handling hashes ourselves!

Reply:
I guess we can carry on this discussion more generally when it gets to implementing #703, for now, @chrisrothUT should we switch to @femtobit's suggestion?

Reply:
Yeah that works. I guess we can leave the dataclasses as is and address that on the next PR?

Reply:
I think so – they work, they add something compared to simple `Permutation`s and we'll probably eliminate them soon, so I wouldn't worry. I hope the tests with the new `float→int` logic will pass, so we can merge this after you change the hashing logic

Reply:
Yeah the new int logic worked and it’s definitely a better solution. I think “always use ints when comparing hashes” might be a new rule of mine.

Reply:
That is generally a good rule for anything that hinges on equality testing or reproducible results in general, integer arithmetics is much more reliable and predictable. (Though this might just show I've been exposed too much to theoretical computer scientists who think floating-point numbers are the work of the devil...)